keps/sig-api-machinery/20190426-crd-defaulting.md

```diff
@@ -0,0 +1,336 @@
+
+```yaml
+properties:
+  foo:
+    type: string
+    default: "abc" 
+```
+
+This KEP proposes to apply these default values during deserialization, in the same way as native resources do. We assume _structural schemas_ as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](/keps/sig-api-machinery/20190425-structural-openapi.md).
+
+This feature starts behind a feature gate `CustomResourceDefaulting`, disabled by default as alpha.
+
+## Motivation
+
+* By far most native Golang based resources implement defaulting. CRDs do not allow that, leading to unnatural, not Kubernetes-like APIs. This is bad for the ecosystem.
+* Mutating Admission Webhooks can be used for defaulting, but:
+  * they are not adequate as it is not possible to set default values of newly added fields on GET because admission is not run in the storage layer when reading from etcd.
+  * webhooks are complex, both from the development/maintenance point of view and due to their non-trivial operational overhead. This makes them a no-go for many "not so ambitiously, professionally developed CRDs", e.g. in in-house enterprise environments.
+* _Structural schemas_ as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](/keps/sig-api-machinery/20190425-structural-openapi.md) make defaulting a low hanging fruit.
+ 
+### Goals
+
+* add CustomResource defaulting at the correct position in the request pipeline 
+* allow definition of defaults via the OpenAPI v3 `default` field.
+
+### Non-Goals
+
+* allow non-constant defaults: native Golang code can of course set defaults which depends on other fields of a JSON object. This is out of scope here and would need some kind of defaulting webhook.
+* native-type declarative defaulting: this KEP is about CRDs. Though it might be desirable to support the same `// +default=<some-json-value>` tag and a mechanism in `k8s.io/apiserver` to evaluate defaults in native, generic registries, this is out-of-scope of the KEP though.
+
+## Proposal
+
+We assume the CRD has _structural schemas_ (as defined in [KEP Vanilla OpenAPI Subset: Structural Schema](/keps/sig-api-machinery/20190425-structural-openapi.md)).
+
+We propose to
+
+1. derive the value-validation-less variant of the structural schema (trivial by definition of _structural schema_) and 
+2. recursively follow the given CustomResource instance and the structural schema, applying defaults where an object field is 
+   * undefined (`_, ok := obj[field]; !ok`)
+   * `nil` if the field not nullable
+   * empty in case of lists and maps, and if nullable is not set.
+  
+   This means in detail for primitive types and no types (i.e. IntOrString or arbitrary JSON) in the structural schema:
+   
+   * `if v, ok := obj[fld]; !ok` => default
+   * `else if !nullable && v == nil` => default
+   
+   and for `array` type in the schema one of these:
+   
+   * `if v, ok := obj[fld]; !ok` => default
+   * `else if !nullable && v == nil` => default
```

Comment:
if I'm reading this correctly, we'd want a nullable field to skip all the checks after the first one, right? something like this?
```
* `if v, ok := obj[fld]; !ok` => default
* `else if nullable` => no default
* `else if v == nil` => default
...
```

same for object

Reply:
@liggitt I read this as being very similar to what Stefan wrote, but harder to read?

Reply:
if the schema specifies the field is nullable, and the current value is `[]`, my example would not default, and stefan's would

I *think* we want to avoid treating `nil` and `[]`/`{}` values differently for defaulting purposes

Reply:
Oh I see what happened here.

Reply:
If it's nil and non-nullable, shouldn't it be an error anyway?

Reply:
Let's think in terms of UX behavior:

```
{
  defaulted-list: [] # Do I want this defaulted? Not sure, I'm specifying a value after all ...
}
---
{
  defaulted-list: null # Do I want this defaulted? Not sure, I'm specificying a value after all ...
}
---
{
  # defaulted-list should certainly be defaulted
}
```

One of the principle we used in apply was that if users specified something, then we assume that it's what they want, and setting something to nil is different from not setting it at all.

Reply:
> If it's nil and non-nullable, shouldn't it be an error anyway?

That's a good point. We could default only if the property was completely absent.

Things I like about only defaulting unspecified properties:
* It's simpler to explain
* It's simpler to implement
* It doesn't mess with user data in any way
* It avoids masking schema errors if they explicitly send `null` for a property that is not nullable and has a default

Opposing considerations:
* The distinction between "null" and "absent" is not easily maintained in all serialization formats (notably protobuf), so if that got lost in a round-trip, we could end up applying defaulting anyway (this isn't a huge concern, given we currently maintain the distinction for custom resources, so I think we have to figure out a way to continue doing so if we ever switch to persist in other formats like proto)
* We would not be able to replicate defaulting rules for built-in resources using this standard in a hypothetical world where built-in types get converted to CRDs (this also isn't a huge concern, given there are tons of defaulting rules for built-in types we couldn't replicate using openapi at all)
* Sending explicit `null` values in some patch formats today removes fields, so there's some prior art for `null == unset` in kube. For example, `kubectl patch ... --type=merge -p '{"spec":{"key":null}}'` removes an existing `key` property, it does not persist a literal `null` value. I'm not sure what server-side apply does here.

Overall, I think the simpler and more intuitive behavior is probably what we want.